src/webrtc/groupCall.ts

```diff
@@ -0,0 +1,565 @@
+            return;
+        }
+
+        // Get the latest member participant state event.
+        const memberStateEvent = this.room.currentState.getStateEvents(
+            "m.room.member",
+            member.userId,
+        );
+        const memberStateContent = memberStateEvent.getContent();
+
+        if (!memberStateContent) {
+            return;
+        }
+
+        const participantInfo = memberStateContent[CONF_PARTICIPANT];
+
+        if (!participantInfo || typeof participantInfo !== "object") {
+            return;
+        }
+
+        const { expiresAt, sessionId } = participantInfo;
+
+        // If the participant state has expired, ignore this user.
+        const now = new Date().getTime();
+
+        if (expiresAt < now) {
+            return;
+        }
+
+        // If there is an existing participant for this member check the session id.
+        // If the session id changed then we can hang up the old call and start a new one.
+        // Otherwise, ignore the member change event because we already have an active participant.
+        let participant = this.participants.find(
+            (p) => p.member.userId === member.userId,
+        );
+
+        if (participant) {
+            if (participant.sessionId !== sessionId) {
+                participant.call.hangup(CallErrorCode.Replaced, false);
+            } else {
+                return;
+            }
+        }
+
+        // Only initiate a call with a user who has a userId that is lexicographically
+        // less than your own. Otherwise, that user will call you.
+        if (member.userId < localUserId) {
+            return;
+        }
+
+        const call = this.client.createCall(this.room.roomId, member.userId);
+
+        let callPromise;
+
+        if (this.type === CallType.Video) {
+            callPromise = call.placeVideoCall();
+        } else {
+            callPromise = call.placeVoiceCall();
+        }
+
+        callPromise.then(() => {
+            if (this.dataChannelsEnabled) {
+                call.createDataChannel("datachannel", this.dataChannelOptions);
+            }
+        });
+
+        if (participant) {
+            participant.replaceCall(call, sessionId);
+        } else {
+            participant = new GroupCallParticipant(
+                this,
+                member,
+                sessionId,
+                call,
+            );
+            // TODO: Should we wait until the call has been answered to push the participant?
+            // Or do we hide the participant until their stream is live?
+            // Does hiding a participant without a stream present a privacy problem because
+            // a participant without a stream can still listen in on other user's streams?
+            this.participants.push(participant);
+            this.reEmitter.reEmit(participant, Object.values(GroupCallParticipantEvent));
+            this.emit(GroupCallEvent.ParticipantsChanged), this.participants;
```

Comment:
This looks weird

Reply:
You're talking about the comment right? I could use some feedback on that.

Reply:
I meant it probably should be like this, right? 

```suggestion
            this.emit(GroupCallEvent.ParticipantsChanged, this.participants);
```

Or is this some syntax, I am not aware of :sweat_smile:

Reply:
Ah, good catch haha :sweat_smile: